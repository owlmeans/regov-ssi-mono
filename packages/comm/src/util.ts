/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { generateKeyPairFromSeed } from '@stablelib/x25519'
import {
  CryptoKey, DIDDocument, DIDHelper, DIDPURPOSE_AGREEMENT, KEYCHAIN_ERROR_NO_KEY, KeyPairToCryptoKeyOptions
} from "@owlmeans/regov-ssi-core"
import { CommKey, COMM_DID_AGREEMENT_KEY_DEFAULT, connectionFieldList, DIDCommConnectMeta, ERROR_COMM_DID_NOKEY, ERROR_COMM_NO_RECIPIENT } from "./types"
import { JWE } from 'did-jwt'


export const cryptoKeyToCommKey = async (
  key: CryptoKey, options?: KeyPairToCryptoKeyOptions
): Promise<CommKey> => {
  if (!key.pk) {
    throw Error(KEYCHAIN_ERROR_NO_KEY)
  }
  const pk = Buffer.from(key.pk, 'utf8').subarray(0, 32)
  const pubKey = generateKeyPairFromSeed(pk).publicKey

  return {
    id: options?.id || key.id || '_noid_',
    pk, pubKey
  }
}

export const filterConnectionFields = (connection: DIDCommConnectMeta): DIDCommConnectMeta => {
  return Object.fromEntries(
    connectionFieldList.map(field => [field, connection[field]])
  ) as unknown as DIDCommConnectMeta
}

export const invertConnection = (connection: DIDCommConnectMeta, channel?: string): DIDCommConnectMeta => {
  const newConnection = filterConnectionFields(JSON.parse(JSON.stringify(connection)))

  if (!newConnection.recipient) {
    throw new Error(ERROR_COMM_NO_RECIPIENT)
  }

  return {
    ...newConnection,
    recipientId: newConnection.sender.id,
    recipient: newConnection.sender,
    sender: newConnection.recipient,
    channel: channel || newConnection.channel
  }
}

export const parseJWE = (data: string) => {
  try {
    return JSON.parse(data) as JWE
  } catch (e) {
  }

  return undefined
}

export const didDocToCommKeyBuilder = (helper: DIDHelper) =>
  async (did: DIDDocument, keyId: string = COMM_DID_AGREEMENT_KEY_DEFAULT): Promise<Partial<CommKey>> => {
    const agreement = helper.expandVerificationMethod(did, DIDPURPOSE_AGREEMENT, keyId)
    if (!agreement.publicKeyBase58) {
      throw new Error(ERROR_COMM_DID_NOKEY)
    }

    return {
      id: did.id,
      pubKey: helper.getCrypto().base58().decode(agreement.publicKeyBase58)
    }
  }