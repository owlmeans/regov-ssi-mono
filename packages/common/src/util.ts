import { MaybeArray, Idish } from "./types"

export const normalizeValue = <Type>(value: MaybeArray<Type>): Type[] => {
  return Array.isArray(value)
    ? value
    : value ? [value] : []
}

export const simplifyValue = <Type>(value: MaybeArray<Type>) => {
  return Array.isArray(value) && value.length < 2
    ? value.length === 0 ? undefined : value[0]
    : value
}

export const singleValue = <Type>(value: MaybeArray<Type>): Type | undefined => {
  return Array.isArray(value)
    ? value.length > 0 ? value[0] : undefined
    : value
}

export const mapValue = async <Type, Res>(
  value: MaybeArray<Type>,
  callback: (value: Type) => Promise<Res>
): Promise<Res[]> => {
  return Promise.all(normalizeValue(value).map(callback))
}

export const addToValue = <Type>(value: MaybeArray<Type> | undefined, add: MaybeArray<Type>): MaybeArray<Type> => {
  if (value) {
    return [...normalizeValue(value), ...normalizeValue(add)]
  }

  return add
}

export const extractId = (id: Idish) => typeof id === 'string' ? id : id.id

export const getDeepValue = <Type, Subject extends {}>(subject: Subject, key: string, def?: Type): Type => {
  const parts = key.split('.')
  return parts.reduce((value, key) => {
    if (subject.hasOwnProperty(key)) {
      const tmp = (subject as any)[key]
      subject = tmp

      return tmp
    }

    return def
  }, def) as Type
}