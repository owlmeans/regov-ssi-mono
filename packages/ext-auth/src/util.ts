/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Credential, Presentation } from "@owlmeans/regov-ssi-core";
import { 
  AuthSubject, REGOV_AUTH_REQUEST_TYPE, REGOV_AUTH_RESPONSE_TYPE, REGOV_CREDENTIAL_TYPE_AUTH 
} from "./types";


export const getAuthFromPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(REGOV_AUTH_RESPONSE_TYPE)
   && !presentation.type.includes(REGOV_AUTH_REQUEST_TYPE)) {
    return undefined
  }

  return presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_CREDENTIAL_TYPE_AUTH)
  )
}

export const getAuthSubject = (cred: Credential): AuthSubject => {
  return cred.credentialSubject as unknown as AuthSubject
}

export const pinValidation = {
  required: true,
  minLength: 4,
  pattern: /^\d+$/
}