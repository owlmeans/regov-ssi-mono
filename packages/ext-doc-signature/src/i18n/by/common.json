{
  "menu": {
    "new": {
      "signature": "Подпіс дакументу"
    },
    "request": {
      "signature": "Запыт подпісу дакуманту"
    }
  },
  "dashboard": {
    "widget": {
      "signature": {
        "add": "Падпісаць дакумент"
      },
      "menu": {
        "title": "Подпіс дакументу",
        "request": "Запытаць подпіс",
        "create": "Стварыць подпіс"
      }
    }
  },
  "signature": {
    "docType": {
      "binary": "Бінарны файл",
      "json": "Дакумент JSON",
      "text": "Просты тэкставы дакумент",
      "unknown": "Невядомы тып файлу"
    },
    "creation": {
      "title": "Пасведчанне подпісу дакумента",
      "file": {
        "mobile": {
          "import": {
            "label": "Нацьсніце тут каб запампаваць файл",
            "loaded": "Дакумент запампаваны."
          },
          "input": {
            "label": "Дакумент",
            "hint": "Калі ласка, дадайце пасведчанне ці прадстаўленне пасведчання дакумента",
            "error": {
              "json": "Прадстаўленны дакумент пашкоджаны ці не з'яўляецца JSON",
              "unknown": "Незнаёмы тып дакументу. Мы не можам яго апрацавць."
            }
          }
        },
        "browse": "Прагляд",
        "binary": "Вы запампавалі бінарны файл. Сістэма не можа адлюстраваць яго змест.",
        "import": {
          "here": "Перацягніце сюды файл ці націсніце, каб пабачыць іншыя мажлівасці.",
          "drop": "Адпусціце файл тут...",
          "loaded": "Дакумент запампаваны."
        },
        "input": {
          "label": "Змест тэкставага дакумента.",
          "hint": "Тут можна змесціць тэкставы дакуманта. Можна карыстацца фарматам JSON."
        }
      },
      "name": {
        "label": "Імя пасведчання дакумента",
        "hint": "Будзе бачна як імя запыту"
      },
      "description": {
        "label": "Апісанне дакументу",
        "hint": "Карысная інфармацыя пра змест гэтага дакументу",
        "copy": "Капіяваць"
      },
      "url": {
        "label": "Адрэса дакуманту",
        "hint": "Інтэрнэт-адрэса па якім знаходзіцца дакумент.",
        "error": {
          "url": "Несапраўдны фармат URL"
        }
      },
      "version": {
        "label": "Версія",
        "hint": "Чытальная версія дакуманту для людзей і кампутараў",
        "error": {
          "pattern": "Памылка. Версія мусіць утрымліваць толькі лічбы, літары, кропкі і мінусы"
        }
      },
      "author": {
        "label": "Аўтар",
        "hint": "Імя аўтара дакументу"
      },
      "authorId": {
        "label": "ID аўтару",
        "hint": "Ідэнтыфікацыйная паслядоўнасць аўтара дакуманту. Магчыма карыстацца размеркваным ID пасведчання."
      },
      "documentHash": {
        "label": "Кантрольная паслядоўнасць",
        "hint": "Выкарыстоўваецца дзеля праверкі подпіса дакуманту"
      },
      "docType": {
        "label": "Тып дакументу",
        "hint": "Тып прадстаўленага дакументу"
      },
      "filename": {
        "label": "Файл",
        "hint": "Назва файлу дакумента."
      },
      "identity": {
        "hint": "Націсніце сюды, каб абраць ідэнтыфікатар дзеля подпіса дакуманту",
        "noItems": "Не знойдзена ідэнтфікатара дзеля подпісу дакумента",
        "pleaseSelect": "Ідэнтыфікатар не абраны, націсніце каб абраць"
      },
      "creationDate": {
        "label": "Дата стварэння",
        "hint": "Дата, калі дакумант быў зменены апошні раз."
      },
      "create": "Стварыць пасведчанне",
      "success": "Дакумент паспяхова падпісаны."
    },
    "request": {
      "title": "Запыт на подпіс дакументу",
      "file": {
        "mobile": {
          "import": {
            "label": "Нацьсніце тут каб запампаваць файл",
            "loaded": "Дакумент запампаваны."
          },
          "input": {
            "label": "Дакумент",
            "hint": "Калі ласка, дадайце пасведчанне ці прадстаўленне пасведчання дакумента",
            "error": {
              "json": "Прадстаўленны дакумент пашкоджаны ці не з'яўляецца JSON",
              "unknown": "Незнаёмы тып дакументу. Мы не можам яго апрацавць."
            }
          }
        },
        "browse": "Паргляд",
        "binary": "Вы запампавалі бінарны файл. Сістэма не можа адлюстраваць яго змест.",
        "import": {
          "here": "Перацягніце сюды файл ці націсніце дзеля дадатковых магчымасцей.",
          "drop": "Адпусціце файл тут...",
          "loaded": "Дакумент запампаваны."
        },
        "input": {
          "label": "Змест тэкставага дакуманту.",
          "hint": "Тут можна змесціць тэкставы дакуманта. Можна карыстацца фарматам JSON."
        }
      },
      "name": {
        "label": "Імя запыту",
        "hint": "Імя запыту будзе бачна толькі вам"
      },
      "description": {
        "label": "Апісанне дакументу",
        "hint": "Карысная інфармацыя аб дакуменце",
        "copy": "Скапіяваць"
      },
      "url": {
        "label": "URL-адрас дакументу",
        "hint": "Інтэрнэт-адрэса па якому знаходзіцца дакумент",
        "error": {
          "url": "Несапраўдны фармат URL"
        }
      },
      "version": {
        "label": "Версія",
        "hint": "Чытальны нумар версіі дакументу дзеля людзей і кампутараў",
        "error": {
          "pattern": "Памылка. Версія павінна змяшчаць толькі лічбы, літары, кропкі і мінусы"
        }
      },
      "author": {
        "label": "Аўтар",
        "hint": "Імя аўтара дакумента"
      },
      "authorId": {
        "label": "ID аўтара",
        "hint": "Ідэнтфікацыйна паслядоўнасць. Можа быць ID пасведчання."
      },
      "documentHash": {
        "label": "Кантрольная паслядоўнасць",
        "hint": "Выкарыстоўваецца каб правяраць подпіс дакументу"
      },
      "create": "Стварыць запыт",
      "success": "Запыт створаны паспяхова.",
      "meta": {
        "title": "Запыт на подпіс дзеля \"{{name}}\""
      },
      "item": {
        "type": "Запыт подпісу"
      },
      "view": "Запыт подпісу",
      "close": "Зачыніць"
    },
    "response": {
      "title": "Прадставіць запытаны подпіс дакумента",
      "vc": {
        "hint": "Націсніце, каб абраць пасведчанне дезля подпісу адказа.",
        "noItems": "Няма пасведчання подпісу дзеля адказу",
        "pleaseSelect": "Пасведчанне не абрана. Нажміце каб абраць адно"
      },
      "description": {
        "label": "Апісанне запытанага дакументу",
        "hint": "Карысная інфармацыя аб дакуменце, каб зразумець як стварыць адказ."
      },
      "url": {
        "label": "URL-адрас дакументу",
        "hint": "Інтэрнэт-адрас па якому знаходзіцца дакумент"
      },
      "version": {
        "label": "Версія",
        "hint": "Чытальны нумар версіі дакументу для людзей і кампутараў"
      },
      "authorId": {
        "label": "ID аўтару",
        "hint": "Ідэнтыфікацыйная паслядоўнасць аўтара."
      },
      "documentHash": {
        "label": "Кантрольная паслядоўнасць",
        "hint": "Выкарыстоўваецца каб праверыць подпіс дакументу"
      },
      "alert": {
        "error": {
          "wrongHash": "Абраны подпіс належыць іншаму дакументу",
          "noRequest": "Прадстаўлены адказ не запытваўся",
          "wrongDocument": "Паданы подпіс належыць іншаму дакументу",
          "noIdentity": "Няма ідэнтыфікатру дзеля подпісу дакументу",
          "unknwonValidationError": "Паданы дакумент не з'яўляецца дапушчальным запытам пасведчання. Прычына наведяома."
        }
      },
      "credential": {
        "title": "Дакумент {{type}} {{hash}}",
        "added": "Пасведчанне паспяхова дададзена ў гэты кашалёк."
      },
      "produce": "Адказаць",
      "produced": "Адказ с подпісам створаны паспяхова",
      "accept": "Прыняць",
      "close": "Зачыніць"
    },
    "view": {
      "title": "Пасведчанне подпісу дакумента",
      "signedAt": {
        "hint": "Дата подпісу",
        "label": "Дата подпісу"
      },
      "description": {
        "label": "Апісанне"
      },
      "documentHash": {
        "label": "Кантрольная паслядоўнасць"
      },
      "filename": {
        "label": "Імя файлу"
      },
      "url": {
        "label": "URL"
      },
      "authorId": {
        "label": "ID аўтара"
      },
      "author": {
        "label": "Аўтар"
      },
      "creationDate": {
        "hint": "Дата стварэння",
        "label": "Дата стварэння"
      },
      "version": {
        "hint": "Версія",
        "label": "Версія"
      },
      "docType": {
        "hint": "Тып дакументу",
        "label": "Тып дакументу"
      },
      "upload": "Загрузіць файл",
      "file": {
        "mobile": {
          "import": {
            "label": "Нацьсніце тут каб запампаваць файл",
            "loaded": "Дакумент запампаваны."
          },
          "input": {
            "label": "Дакумент",
            "hint": "Калі ласка, дадайце пасведчанне ці прадстаўленне пасведчання дакумента",
            "error": {
              "json": "Прадстаўленны дакумент пашкоджаны ці не з'яўляецца JSON",
              "unknown": "Незнаёмы тып дакументу. Мы не можам яго апрацавць."
            }
          }
        },
        "browse": "Адчыніць...",
        "input": {
          "hint": "Змест дакуманту ці JSON дзеля праверкі.",
          "label": "Тэкст дакумента"
        },
        "import": {
          "here": "Перацягніце сюды файл, каб даведацца ці ён падпісаны гэтым подпісам.",
          "drop": "Адпусціце файл тут...",
          "loaded": "Дакумент запампаваны."
        }
      },
      "alert": {
        "error": {
          "matches": "Прадстаўлены дакумент падпісаны гэтым подпісам.",
          "notmatch": "Прадстаўлены дакумент не падпісаны гэтым подпісам.",
          "severity": {
            "error": "error",
            "success": "success"
          },
          "title": {
            "error": "Памылка",
            "success": "Поспех"
          }
        }
      }
    },
    "list": {
      "item": {
        "unknown": "Невядомы подпіс дакуманта",
        "type": "Пасведчанне подпісу дакумента"
      }
    }
  }
}