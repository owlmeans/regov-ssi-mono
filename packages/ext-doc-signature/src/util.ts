import { Presentation } from "@owlmeans/regov-ssi-core"
import { REGOV_SIGNATURE_REQUEST_TYPE } from "./types"


export const getSignatureRequestFromPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(REGOV_SIGNATURE_REQUEST_TYPE)) {
    return undefined
  }

  return presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_SIGNATURE_REQUEST_TYPE)
  )
}