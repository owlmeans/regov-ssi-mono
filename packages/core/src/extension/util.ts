/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, normalizeValue } from "../common"
import { BasicCredentialType, Credential } from "../vc"
import {
  CredentialService, CredentialServiceBuilder, defaultBuildMethod, defaultClaimMethod,
  defaultOfferMethod, defaultRequestMethod, defaultRespondMethod, defaultSignMethod, defaultValidateMethod,
  Extension,
  ValidationResult
} from "./ext"
import { CredentialDescription } from "./schema"


export const findAppropriateCredentialType = (
  ext: Extension, types: BasicCredentialType, defaultType: string
): string => {
  types = normalizeValue(types)

  return (Object.entries(ext.factories).map(([type]) => type)
    .find(type => types.includes(type)) || defaultType)
}

const methodMap = {
  'produceBuildMethod': 'build',
  'produceSignMethod': 'sign',
  'produceValidateMethod': 'validate',
  'produceClaimMethod': 'claim',
  'produceOfferMethod': 'offer',
  'produceRequestMethod': 'request',
  'produceRespondMethod': 'respond'
}

export const addFactoriesToExt = (ext: Extension, type: string, factories: CredentialServiceBuilder) => {
  const description = (ext.schema.credentials as { [key: string]: CredentialDescription })[type]
  ext.factories[type] = {
    ...{
      build: defaultBuildMethod(description),
      sign: defaultSignMethod(description),
      validate: defaultValidateMethod(description),
      claim: defaultClaimMethod(description),
      offer: defaultOfferMethod(description),
      request: defaultRequestMethod(description),
      respond: defaultRespondMethod(description)
    },
    ...Object.entries(factories).reduce((_facts, [method, builder]) => {
      if (ext.schema.credentials) {
        return {
          ..._facts,
          [(methodMap as any)[method]]: builder(description)
        }
      }
      return _facts
    }, {} as CredentialService)
  }
}

interface CredentialChainDebug {
  id: string
  type: string[]
  children: CredentialChainDebug[]
}

export const debugCredentialChain = (credential: Credential): CredentialChainDebug => {
  const result: CredentialChainDebug = { id: credential.id, type: credential.type, children: [] }

  result.children = normalizeValue(credential.evidence).map(
    evidence => debugCredentialChain(evidence as Credential)
  )

  return result
}

export const debugValidationResult = (result: ValidationResult): CredentialChainDebug => {
  const _result: CredentialChainDebug = {
    id: result.instance?.id || '',
    type: result.instance?.type || [''],
    children: []
  }

  _result.children = normalizeValue(result.evidence).map(
    evidence => debugValidationResult(evidence.result)
  )

  return _result
}