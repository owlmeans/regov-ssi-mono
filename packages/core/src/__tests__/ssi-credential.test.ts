/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

require('dotenv').config()

import { nodeCryptoHelper } from "../common"
import {
  buildDidHelper, buildDidRegistryWarpper, DIDDocument, DIDPURPOSE_ASSERTION, DIDPURPOSE_AUTHENTICATION,
  DIDPURPOSE_VERIFICATION,
  VERIFICATION_KEY_HOLDER
} from "../did"
import {
  Presentation, UnsignedPresentation, buildSSICore, SSICore, Credential, UnsignedCredential, buildKeyChain
} from "../index"

import util from 'util'
util.inspect.defaultOptions.depth = 8


const test: {
  ssi?: SSICore
  unsigned?: UnsignedCredential<{}>
  credential?: Credential
  unsignedP?: UnsignedPresentation
  presentation?: Presentation
} = {}

beforeAll(async () => {
  test.ssi = await buildSSICore({
    keys: await buildKeyChain({
      password: '11111111',
      crypto: nodeCryptoHelper
    }),
    crypto: nodeCryptoHelper,
    did: buildDidRegistryWarpper(buildDidHelper(nodeCryptoHelper))
  })
})

describe('SSI Verifiable Credential', () => {
  it('creates Credentail', async () => {
    if (!test.ssi) {
      throw 'Setup didn\'t provide SSICore'
    }

    const subject = {
      data: {
        '@type': 'TestCredentialSubjectDataType',
        worker: 'Valentin Michalych'
      }
    }

    const key = await test.ssi.keys.getCryptoKey()
    const didUnsigned = await test.ssi.did.helper().createDID(
      key,
      {
        data: JSON.stringify(subject),
        hash: true,
        purpose: [DIDPURPOSE_VERIFICATION, DIDPURPOSE_ASSERTION, DIDPURPOSE_AUTHENTICATION]
      }
    )

    const did = await test.ssi.did.helper().signDID(key, didUnsigned)
    expect(await test.ssi.did.helper().verifyDID(did)).toBe(true)

    test.ssi.did.addDID(did)

    const unsingnedCredentail = await test.ssi.buildCredential({
      id: did.id,
      type: ['VerifiableCredential', 'TestCredential'],
      holder: did,
      context: {
        '@version': 1.1,
        xsd: 'http://www.w3.org/2001/XMLSchema#',
        exam: 'https://example.org/vc-schema#',
        data: {
          '@id': 'exam:data',
          '@type': '@id',
          '@context': {
            worker: { '@id': 'exam:worker', '@type': 'xsd:string' }
          }
        }
      },
      subject
    })

    test.unsigned = unsingnedCredentail
    expect(unsingnedCredentail).toMatchSnapshot({
      id: expect.any(String),
      holder: expect.any(Object),
      issuanceDate: expect.any(String)
    })
  })

  it('signs Credential', async () => {
    if (!test.ssi) {
      throw 'Setup didn\'t provide SSICore'
    }
    if (!test.unsigned) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }

    const did = <DIDDocument>await test.ssi.did.lookUpDid(test.unsigned.id)
    const credentail = await test.ssi.signCredential(
      test.unsigned,
      did,
      { keyId: VERIFICATION_KEY_HOLDER }
    )

    test.credential = credentail

    expect(credentail).toMatchSnapshot({
      id: expect.any(String),
      holder: expect.any(Object),
      issuer: expect.any(Object),
      issuanceDate: expect.any(String),
      proof: {
        created: expect.any(String),
        jws: expect.any(String),
        verificationMethod: expect.any(String),
      }
    })
  })

  it('verifies Credential', async () => {
    if (!test.ssi) {
      throw 'Setup didn\'t provide SSICore'
    }
    if (!test.credential) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }
    const [result, _] = await test.ssi.verifyCredential(test.credential)

    expect(result).toBe(true)
  })

  it('doesn\'t allow to temper subject', async () => {
    if (!test.ssi) {
      throw 'Setup didn\'t provide SSICore'
    }
    if (!test.credential) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }

    const newCredential = <Credential<{
      data: {
        '@type': string,
        worker: string
      }
    }>>{ ...test.credential }
    newCredential.credentialSubject = { ...newCredential.credentialSubject }
    newCredential.credentialSubject.data = { ...newCredential.credentialSubject.data }

    newCredential.credentialSubject.data.worker = `${newCredential.credentialSubject.data.worker}_`

    const [result] = await test.ssi.verifyCredential(newCredential)

    expect(result).toBe(false)
  })

  it('creates unsigned verifiable Presentation', async () => {
    if (!test.ssi) {
      throw 'Setup didn\'t provide SSICore'
    }
    if (!test.credential) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }

    const did = <DIDDocument>await test.ssi.did.lookUpDid((test.credential.holder as DIDDocument).id)

    const vp = await test.ssi?.buildPresentation(
      [test.credential],
      { id: did.id, holder: did, type: 'TestPresentation' }
    )

    test.unsignedP = vp

    expect(vp).toMatchSnapshot(
      {
        id: expect.any(String),
        holder: expect.any(Object),
        verifiableCredential: [
          {
            holder: expect.any(Object),
            id: expect.any(String),
            issuanceDate: expect.any(String),
            issuer: expect.any(Object),
            proof: {
              created: expect.any(String),
              jws: expect.any(String),
              verificationMethod: expect.any(String),
            }
          },
        ]
      }
    )
  })

  it('signs verifiable Presentation', async () => {
    if (!test.ssi) {
      throw 'Setup didn\'t provide SSICore'
    }
    if (!test.unsignedP) {
      throw 'Previous test didn\'t provide UnsingedPresentation'
    }
    const did = await test.ssi.did.lookUpDid<DIDDocument>((test.unsignedP.holder as DIDDocument).id)
    if (!did) {
      throw 'No related did in registry'
    }

    const vp = await test.ssi.signPresentation(test.unsignedP, did)

    test.presentation = vp

    expect(vp).toMatchSnapshot({
      id: expect.any(String),
      holder: expect.any(Object),
      verifiableCredential: [
        {
          holder: expect.any(Object),
          id: expect.any(String),
          issuanceDate: expect.any(String),
          issuer: expect.any(Object),
          proof: {
            created: expect.any(String),
            jws: expect.any(String),
            verificationMethod: expect.any(String),
          }
        },
      ],
      proof: {
        jws: expect.any(String),
        verificationMethod: expect.any(String),
        challenge: expect.any(String),
        domain: expect.any(String),
        created: expect.any(String)
      }
    })
  })

  it('verifies verifiable Presentation', async () => {
    if (!test.ssi) {
      throw 'Setup didn\'t provide SSICore'
    }
    if (!test.presentation) {
      throw 'Previous test didn\'t provide Presentation'
    }
    if (!test.presentation.id) {
      throw 'Previous test didn\'t provide Presentation with id'
    }

    const [result, _] = await test.ssi.verifyPresentation(test.presentation)

    expect(result).toBe(true)
  })
})