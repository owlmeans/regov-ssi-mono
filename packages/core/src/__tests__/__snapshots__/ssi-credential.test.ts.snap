// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSI Verifiable Credential creates Credentail 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
    Object {
      "@version": 1.1,
      "data": Object {
        "@context": Object {
          "worker": Object {
            "@id": "exam:worker",
            "@type": "xsd:string",
          },
        },
        "@id": "exam:data",
        "@type": "@id",
      },
      "exam": "https://example.org/vc-schema#",
      "xsd": "http://www.w3.org/2001/XMLSchema#",
    },
  ],
  "credentialSubject": Object {
    "data": Object {
      "@type": "TestCredentialSubjectDataType",
      "worker": "Valentin Michalych",
    },
  },
  "holder": Any<Object>,
  "id": Any<String>,
  "issuanceDate": Any<String>,
  "type": Array [
    "VerifiableCredential",
    "TestCredential",
  ],
}
`;

exports[`SSI Verifiable Credential creates unsigned verifiable Presentation 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
  ],
  "holder": Any<Object>,
  "id": Any<String>,
  "type": Array [
    "VerifiablePresentation",
    "TestPresentation",
  ],
  "verifiableCredential": Array [
    Object {
      "@context": Array [
        "https://www.w3.org/2018/credentials/v1",
        Object {
          "@version": 1.1,
          "data": Object {
            "@context": Object {
              "worker": Object {
                "@id": "exam:worker",
                "@type": "xsd:string",
              },
            },
            "@id": "exam:data",
            "@type": "@id",
          },
          "exam": "https://example.org/vc-schema#",
          "xsd": "http://www.w3.org/2001/XMLSchema#",
        },
      ],
      "credentialSubject": Object {
        "data": Object {
          "@type": "TestCredentialSubjectDataType",
          "worker": "Valentin Michalych",
        },
      },
      "holder": Any<Object>,
      "id": Any<String>,
      "issuanceDate": Any<String>,
      "issuer": Any<Object>,
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "assertionMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "type": Array [
        "VerifiableCredential",
        "TestCredential",
      ],
    },
  ],
}
`;

exports[`SSI Verifiable Credential signs Credential 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
    Object {
      "@version": 1.1,
      "data": Object {
        "@context": Object {
          "worker": Object {
            "@id": "exam:worker",
            "@type": "xsd:string",
          },
        },
        "@id": "exam:data",
        "@type": "@id",
      },
      "exam": "https://example.org/vc-schema#",
      "xsd": "http://www.w3.org/2001/XMLSchema#",
    },
  ],
  "credentialSubject": Object {
    "data": Object {
      "@type": "TestCredentialSubjectDataType",
      "worker": "Valentin Michalych",
    },
  },
  "holder": Any<Object>,
  "id": Any<String>,
  "issuanceDate": Any<String>,
  "issuer": Any<Object>,
  "proof": Object {
    "created": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "assertionMethod",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "type": Array [
    "VerifiableCredential",
    "TestCredential",
  ],
}
`;

exports[`SSI Verifiable Credential signs verifiable Presentation 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
  ],
  "holder": Any<Object>,
  "id": Any<String>,
  "proof": Object {
    "challenge": Any<String>,
    "created": Any<String>,
    "domain": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "authentication",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "type": Array [
    "VerifiablePresentation",
    "TestPresentation",
  ],
  "verifiableCredential": Array [
    Object {
      "@context": Array [
        "https://www.w3.org/2018/credentials/v1",
        Object {
          "@version": 1.1,
          "data": Object {
            "@context": Object {
              "worker": Object {
                "@id": "exam:worker",
                "@type": "xsd:string",
              },
            },
            "@id": "exam:data",
            "@type": "@id",
          },
          "exam": "https://example.org/vc-schema#",
          "xsd": "http://www.w3.org/2001/XMLSchema#",
        },
      ],
      "credentialSubject": Object {
        "data": Object {
          "@type": "TestCredentialSubjectDataType",
          "worker": "Valentin Michalych",
        },
      },
      "holder": Any<Object>,
      "id": Any<String>,
      "issuanceDate": Any<String>,
      "issuer": Any<Object>,
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "assertionMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "type": Array [
        "VerifiableCredential",
        "TestCredential",
      ],
    },
  ],
}
`;
